[browser][websockets][tests] Fix WebSocket Abort tests multiple failures

kjpou1 opened the pull request:
Multiple fixes for Abort.

There are multiple failures with the Abort tests causing some tests to:

- Never end
- Multiple GC calls that end in running out of memory. 

- Some failures are expected to fail with specific WebSocketError codes:

Error codes that are not correct tend to be when websockets tests are looking for specific error codes and messages when the browser returns `NativeError` and the websocket native error code instead.

```
failed: WebSocket is closed before the connection is established.
```

---

These need to be marked correctly and or caught and wrapped correctly.

```
Assert.Equal() Failure
  info: Expected: Aborted
  info: Actual:   Closed
```

Fixes for these should also fix some, note not all, of the memory runtime errors as per issue: https://github.com/dotnet/runtime/issues/45586

--- 

Fix tests that were failing when expecting CloseSent as a valid state.  

```
// fail: [FAIL] System.Net.WebSockets.Client.Tests(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Throws() Failure
//   info: Expected: typeof(System.OperationCanceledException)
//   info: Actual:   typeof(System.Net.WebSockets.WebSocketException): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseSent'
```

---

Tests are intermittently failing with System.OperationCanceledException : The operation was canceled.

~~Added ActiveIssue for these two tests.  https://github.com/dotnet/runtime/issues/46909~~

```
It seems that sometimes the default timeout is not enough for browser so it is extended for browser environments
See issue https://github.com/dotnet/runtime/issues/46909
```

Further addressed in https://github.com/dotnet/runtime/pull/47906


---

SendReceive_PartialMessageBeforeCompleteMessageArrives_Success which describes the test as Ask the remote server to echo back received messages without ever signaling "end of message".  This test hangs as javascript fetch does not support this functionality.

See issue https://github.com/dotnet/runtime/issues/46983 for more detailed description and native javascript tests.

---

Associate PR's: https://github.com/dotnet/runtime/issues/44666 and https://github.com/dotnet/runtime/pull/45470

Close https://github.com/dotnet/runtime/issues/45586
Fix https://github.com/dotnet/runtime/issues/45674

kjpou1 committed:
Set the WebSocketException to Faulted to fix tests

kjpou1 committed:
Fix the exceptions that are being thrown to coincide with test expectations

kjpou1 committed:
Fix Dispose we are not processing it multiple times.

kjpou1 committed:
Fix Abort code so the messages align correctly with the tests.

example from tests:

```
Assert.Equal() Failure
  info: Expected: Aborted
  info: Actual:   Closed
```

kjpou1 committed:
Set the WebSocketException to Faulted to fix test expectations

kjpou1 committed:
Close the connections correctly.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix Abort test Abort_CloseAndAbort_Success

- This was causing a never ending Task when aborted after a Close already executed.
- Never ended which was a cause of memory errors after left running.
- See also https://github.com/dotnet/runtime/issues/45586

kjpou1 committed:
- Fixes for ReceiveAsyncTest

- Exception text not sent correctly.  This test was expecting 'Aborted'.
- Mismatched expected exception messages
- Make sure the connection is torn down.
- Multiple GC calls that end in running out of memory fixed.  https://github.com/dotnet/runtime/issues/45586

```
//  [FAIL] System.Net.WebSockets.Client.Tests.CancelTest.ReceiveAsync_AfterCancellationDoReceiveAsync_ThrowsWebSocketException(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Equal() Failure
//   info:                                  ↓ (pos 39)
//   info: Expected: ··· an invalid state ('Aborted') for this operation. Valid state···
//   info: Actual:   ··· an invalid state ('Open') for this operation. Valid states a···
//   info:                                  ↑ (pos 39)

```

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -318,23 +323,27 @@ private void NativeCleanup()
-                // No cleanup required.
-                return;
-            }
+                if (_state < (int)InternalState.Aborted) {
+                    Interlocked.Exchange(ref _state, (int)InternalState.Disposed);
```
I'm curious about the Interlocked. If relevant code can be running on multiple threads concurrently, then what's to prevent the _state < Aborted being invalid the moment after it's checked?  And if relevant code can't be running on multiple threads concurrently, why the interlocked?

kjpou1 replied:
Will take a look at these.  Thanks for the heads up.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

# Conflicts:
#	src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

lewing committed:
Remove ActiveIssue attributes that should be fixed

lewing commented:
/azp run runtime-libraries-mono outerloop

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add back code after merge conflict

- Update tests that are resolved.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix tests that were failing when expecting CloseSent as a valid state.

```
// fail: [FAIL] System.Net.WebSockets.Client.Tests(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Throws() Failure
//   info: Expected: typeof(System.OperationCanceledException)
//   info: Actual:   typeof(System.Net.WebSockets.WebSocketException): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseSent'
```

kjpou1 committed:
Fix the Abort and Cancel never ending tests.

- Fix the clean up of the task and set or cancel the task cleanly.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

- intermittently failing with System.OperationCanceledException : The operation was canceled.

kjpou1 committed:
Add extra time to timeout for a couple of tests that were intermittently failing with System.OperationCanceledException : The operation was canceled.

- This was an ActiveIssue https://github.com/dotnet/runtime/issues/46909 but extending the time seems to do the job.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

- [browser][websocket] Hang with responses without ever signaling "end of message"
- [ActiveIssue("https://github.com/dotnet/runtime/issues/46983", TestPlatforms.Browser)]  // JS Fetch does not support see issue

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -112,15 +118,17 @@ readyState switch
-            int priorState = Interlocked.CompareExchange(ref _state, (int)InternalState.Connecting, (int)InternalState.Created);
-            if (priorState == (int)InternalState.Disposed)
+            // Check that we have not started already.
+            switch ((InternalState)Interlocked.CompareExchange(ref _state, (int)InternalState.Connecting, (int)InternalState.Created))
```
I may have asked this before, but... why is CompareExchange being used?  I thought this component was entirely single threaded?

kjpou1 replied:
I actually had this in another branch and PR but ran out of time.  Went ahead and addressed this here.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Remove `Interlocked` code as per review comment.

kjpou1 committed:
Fix comment

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix Abort tests on non chrome browsers.

- Fix for those browsers that do not set Close and send an onClose event in certain instances i.e. firefox and safari.
- Chrome will send an onClose event and we tear down the websocket there.  Other browsers need to be handled differently.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

karelz commented on src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:
```diff
@@ -329,7 +328,10 @@ public async Task SendReceive_VaryingLengthBuffers_Success(Uri server)
+
+                // It seems that sometimes the default timeout is not enough for browser so we will extend it
+                // See issue https://github.com/dotnet/runtime/issues/46909
+                var ctsDefault = PlatformDetection.IsBrowser ? new CancellationTokenSource(BrowserTimeOutMilliseconds) : new CancellationTokenSource(TimeOutMilliseconds);
```
Is there harm in merging the 2 timeouts and use 30s also for non-Browser?

kjpou1 replied:
Just did not want change existing functionality.

kjpou1 replied:
Can we do this in a follow up PR?  Really waiting for this to be approved as have another PR waiting on this one.

kjpou1 replied:
addressed in https://github.com/dotnet/runtime/pull/47906

karelz commented:
@kjpou1 any plans for this PR? There doesn't seem to be update in last 17 days ...

kjpou1 commented:
@karelz Waiting for approval.  We can ping some others @marek-safar @lewing @radical @stephentoub

lewing commented:
/azp run runtime-libraries-mono outerloop

kjpou1 commented:
The errors in the outerloop are for the following:

```
error : Work item System.Memory.Tests
error : Work item System.Text.Json.Tests
error : Work item System.Threading.Tasks.Parallel.Tests
error : Work item System.Threading.Tasks.Tests 
```

kjpou1 commented:
Closed by  which was branched off of this PR: https://github.com/dotnet/runtime/pull/47906
